whitepapers/Dates.md

```diff
@@ -0,0 +1,72 @@
+
+### Goals
+Allow storing and displaying works with partially-known publication dates.
+
+## Changes
+
+Create a custom `PartialDate` class with:
+* Three main properties:
+    * `year`: integer (required)
+    * `month`: integer (optional)
+    * `day`: integer (optional)
+* Missing/null values indicate an unknown date part
+* Validation that set values are the correct type and correspond to a valid date
+    * If `month` is missing, `day` must also be missing
+* A way to construct a `PartialDate` from a `Date`-like object, string, or 8-digit integer (see below)
+
+### Models
+* Custom Django field `PartialDateField` that inherits from `IntegerField`
+    * Store a `PartialDate` instance as an 8-digit integer in the database, `YYYYMMDD`
+    * If `month` or `day` is null, put `00` in the `MM` or `DD` segment, respectively
+        * "February 26, 2016" stored as `20160226`
+        * "February 2016" stored as `20160200`
+        * "2016" stored as `20160000`
+    * This allows sorting and filtering by a `PartialDateField` in the database
+        * Values with unknown month and day are sorted at the beginning of the year
+        * Values with known month but unknown day are sorted at the beginning of the month
+* Make `AbstractCreativeWork.date_published` a nullable `PartialDateField`
+
+### Migrations
+* Change the `date_published` columns on `share_creativework` and `share_creativeworkversion` from `DATETIME` to `INTEGER`
+* Convert datetime values into integers, as above, dropping the "time" part of the datetime
+    * If a value is at exactly midnight on the first day of the month, assume unknown day
+        * If the month is January, also assume unknown month
+    * Examples:
+        * `2016-03-20 06:02:11` becomes `20160320`
+        * `2016-03-20 00:00:00` becomes `20160320`
+        * `2016-03-01 06:02:11` becomes `20160301`
+        * `2016-03-01 00:00:00` becomes `20160300`
+        * `2016-01-01 06:02:11` becomes `20160101`
+        * `2016-01-01 00:00:00` becomes `20160000`
+
+### Transformers
+* Add a `PartialDateParser` Transformer link that parses a string into a `PartialDate` instance
+    * Should also handle dates split into an array (e.g. `[2016, 3]`)
+* Use `PartialDateParser` instead of `DateParser` for `CreativeWork.date_published` field in all Transformers
+
+### JSON APIs
+* Serialize `PartialDateField`s in JSON as an array of 3 date parts:
+    * `20160101` serialized as `[2016, 1, 1]`
+    * `20160100` serialized as `[2016, 1, null]`
+    * `20160000` serialized as `[2016, null, null]`
+
+### Search Index
+* Change `date_published` mapping to accept multiple date formats
+    * `'format': 'yyyy-MM-dd||yyyy-MM||yyyy'`
+* Update elasticsearch bot to format `date_published` accordingly
+    * `20160101` indexed as "2016-01-01"
+    * `20160100` indexed as "2016-01"
+    * `20160000` indexed as "2016"
+* Add unindexed `date_published.date_parts` field that contains an array of date parts, as in the JSON APIs
```

Comment:
If `date_published` were an object field, we couldn't query by `date_published` directly. We'd have to put the date in a subfield like `date_published.date`, which isn't good. Now that I'm thinking about it, `date_published.date_parts` would only work on an object field, which also isn't good...

The point of the `date_parts` field was to make the partial date available in search results in a format consistent with the rest of the API. `_source.date_published` will have the original string form (e.g. "2017-03"). We could put the parts in another field (`date_published_parts`?), either as an array (for consistence, would probably be unindexed) or an object (indexed for ease of querying).

Another option is to use [multi-fields](https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-fields.html). We could use custom analyzers with a [pattern tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-pattern-tokenizer.html) to pull the year/month/day into `date_published.year`, etc. They would be queryable but wouldn't show up in the search results. If the string form is sufficient to display search results (it could even be shown as-is), this would give the nicest query interface.

Thoughts?